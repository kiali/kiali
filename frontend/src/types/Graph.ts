import { Namespace } from './Namespace';
import { AppenderString, DurationInSeconds, TimeInSeconds } from './Common';
import { Health } from './Health';
import { HealthAnnotationType } from './HealthAnnotation';

export interface Layout {
  name: string;
}

export const SUMMARY_PANEL_CHART_WIDTH = 250;

export type SummaryType = 'graph' | 'node' | 'edge' | 'box';

export interface SummaryData {
  isPF?: boolean;
  summaryTarget: any;
  summaryType: SummaryType;
}

export enum Protocol {
  GRPC = 'grpc',
  HTTP = 'http',
  TCP = 'tcp'
}

export interface SummaryPanelPropType {
  data: SummaryData;
  duration: DurationInSeconds;
  graphType: GraphType;
  injectServiceNodes: boolean;
  kiosk: string;
  namespaces: Namespace[];
  queryTime: TimeInSeconds;
  rateInterval: string;
  step: number;
  trafficRates: TrafficRate[];
}

export enum EdgeMode {
  ALL = 'all',
  NONE = 'none',
  UNHEALTHY = 'unhealthy'
}

export enum EdgeLabelMode {
  RESPONSE_TIME_AVERAGE = 'avg',
  RESPONSE_TIME_GROUP = 'responseTime',
  RESPONSE_TIME_P50 = 'rt50',
  RESPONSE_TIME_P95 = 'rt95',
  RESPONSE_TIME_P99 = 'rt99',
  THROUGHPUT_GROUP = 'throughput',
  THROUGHPUT_REQUEST = 'throughputRequest',
  THROUGHPUT_RESPONSE = 'throughputResponse',
  TRAFFIC_DISTRIBUTION = 'trafficDistribution',
  TRAFFIC_RATE = 'trafficRate'
}

export const isResponseTimeMode = (mode: EdgeLabelMode): boolean => {
  return (
    mode === EdgeLabelMode.RESPONSE_TIME_GROUP ||
    mode === EdgeLabelMode.RESPONSE_TIME_AVERAGE ||
    mode === EdgeLabelMode.RESPONSE_TIME_P50 ||
    mode === EdgeLabelMode.RESPONSE_TIME_P95 ||
    mode === EdgeLabelMode.RESPONSE_TIME_P99
  );
};

export const isThroughputMode = (mode: EdgeLabelMode): boolean => {
  return (
    mode === EdgeLabelMode.THROUGHPUT_GROUP ||
    mode === EdgeLabelMode.THROUGHPUT_REQUEST ||
    mode === EdgeLabelMode.THROUGHPUT_RESPONSE
  );
};

export enum RankMode {
  RANK_BY_INBOUND_EDGES = 'inboundEdges',
  RANK_BY_OUTBOUND_EDGES = 'outboundEdges'
}

export type RankResult = {
  // Number of discrete rankings, N for the current scoring. N in [0..100]. 0 indicates no active rankings.
  upperBound: number;
};

export const numLabels = (modes: EdgeLabelMode[]): number => {
  return modes.filter(m => m !== EdgeLabelMode.RESPONSE_TIME_GROUP && m !== EdgeLabelMode.THROUGHPUT_GROUP).length;
};

export enum TrafficRate {
  GRPC_GROUP = 'grpc',
  GRPC_RECEIVED = 'grpcReceived', // response_messages
  GRPC_REQUEST = 'grpcRequest',
  GRPC_SENT = 'grpcSent', // request_messages
  GRPC_TOTAL = 'grpcTotal', // sent_bytes + received_bytes
  HTTP_GROUP = 'http',
  HTTP_REQUEST = 'httpRequest',
  TCP_GROUP = 'tcp',
  TCP_RECEIVED = 'tcpReceived', // received_bytes
  TCP_SENT = 'tcpSent', // sent_bytes
  TCP_TOTAL = 'tcpTotal' // sent_bytes + received_bytes
}

export const DefaultTrafficRates: TrafficRate[] = [
  TrafficRate.GRPC_GROUP,
  TrafficRate.GRPC_REQUEST,
  TrafficRate.HTTP_GROUP,
  TrafficRate.HTTP_REQUEST,
  TrafficRate.TCP_GROUP,
  TrafficRate.TCP_SENT
];

export const isGrpcRate = (rate: TrafficRate): boolean => {
  return (
    rate === TrafficRate.GRPC_GROUP ||
    rate === TrafficRate.GRPC_RECEIVED ||
    rate === TrafficRate.GRPC_REQUEST ||
    rate === TrafficRate.GRPC_SENT ||
    rate === TrafficRate.GRPC_TOTAL
  );
};

export const toGrpcRate = (rate: string): TrafficRate | undefined => {
  switch (rate) {
    case 'received':
      return TrafficRate.GRPC_RECEIVED;
    case 'requests':
    case 'request':
      return TrafficRate.GRPC_REQUEST;
    case 'sent':
      return TrafficRate.GRPC_SENT;
    case 'total':
      return TrafficRate.GRPC_TOTAL;
    default:
      return undefined;
  }
};

export const isHttpRate = (rate: TrafficRate): boolean => {
  return rate === TrafficRate.HTTP_GROUP || rate === TrafficRate.HTTP_REQUEST;
};

export const toHttpRate = (rate: string): TrafficRate | undefined => {
  switch (rate) {
    case 'requests':
    case 'request':
      return TrafficRate.HTTP_REQUEST;
    default:
      return undefined;
  }
};

export const isTcpRate = (rate: TrafficRate): boolean => {
  return (
    rate === TrafficRate.TCP_GROUP ||
    rate === TrafficRate.TCP_RECEIVED ||
    rate === TrafficRate.TCP_SENT ||
    rate === TrafficRate.TCP_TOTAL
  );
};

export const toTcpRate = (rate: string): TrafficRate | undefined => {
  switch (rate) {
    case 'received':
      return TrafficRate.TCP_RECEIVED;
    case 'sent':
      return TrafficRate.TCP_SENT;
    case 'total':
      return TrafficRate.TCP_TOTAL;
    default:
      return undefined;
  }
};

export enum GraphType {
  APP = 'app',
  SERVICE = 'service',
  VERSIONED_APP = 'versionedApp',
  WORKLOAD = 'workload'
}

export enum BoxByType {
  APP = 'app',
  CLUSTER = 'cluster',
  DATAPLANE = 'dataplane',
  NAMESPACE = 'namespace'
}

export enum NodeType {
  AGGREGATE = 'aggregate',
  APP = 'app',
  BOX = 'box',
  SERVICE = 'service',
  UNKNOWN = 'unknown',
  WORKLOAD = 'workload'
}

export const CLUSTER_DEFAULT = 'Kubernetes'; // Istio default cluster, typically indicates a single-cluster env
export const UNKNOWN = 'unknown';

export interface NodeParamsType {
  aggregate?: string;
  aggregateValue?: string;
  app: string;
  cluster?: string;
  namespace: Namespace;
  nodeType: NodeType;
  service: string;
  version?: string;
  workload: string;
}

// This data is stored in the _global scratch area in the cy graph
// for use by code that needs access to it.
// We can add more props to this scratch data as the need arises.
export const CytoscapeGlobalScratchNamespace = '_global';
export type CytoscapeGlobalScratchData = {
  activeNamespaces: Namespace[];
  edgeLabels: EdgeLabelMode[];
  forceLabels: boolean;
  graphType: GraphType;
  homeCluster: string;
  showOutOfMesh: boolean;
  showSecurity: boolean;
  showVirtualServices: boolean;
  trafficRates: TrafficRate[];
};

export interface CytoscapeBaseEvent {
  summaryTarget: any; // the cytoscape element that was the target of the event
  summaryType: SummaryType; // what the summary panel should show
}

export interface GraphEvent extends CytoscapeBaseEvent {
  isPF?: boolean;
}

// Graph Structures
type PercentageOfTrafficByFlag = {
  [flag: string]: string;
};

type PercentageOfTrafficByHost = {
  [host: string]: string;
};

export type ResponseDetail = {
  flags: PercentageOfTrafficByFlag;
  hosts: PercentageOfTrafficByHost;
};

export type Responses = {
  [responseCode: string]: ResponseDetail;
};

type ValidProtocols = 'http' | 'grpc' | 'tcp';

export type ProtocolNoTraffic = {
  protocol: ValidProtocols;
};

export type ProtocolTrafficHttp = {
  protocol: 'http';
  rates: {
    http: string;
    httpPercentErr?: string;
  };
  responses: Responses;
};

export type ProtocolTrafficGrpc = {
  protocol: 'grpc';
  rates: {
    grpc: string;
    grpcPercentErr?: string;
  };
  responses: Responses;
};

export type ProtocolTrafficTcp = {
  protocol: 'tcp';
  rates: {
    tcp: string;
  };
  responses: Responses;
};

export type ProtocolWithTraffic = ProtocolTrafficHttp | ProtocolTrafficTcp | ProtocolTrafficGrpc;
export type ProtocolTraffic = ProtocolWithTraffic | ProtocolNoTraffic;

export const hasProtocolTraffic = (protocolTraffic: ProtocolTraffic): protocolTraffic is ProtocolWithTraffic => {
  return (
    (protocolTraffic as ProtocolWithTraffic).rates !== undefined &&
    (protocolTraffic as ProtocolWithTraffic).responses !== undefined
  );
};

export const prettyProtocol = (protocol: ValidProtocols): string => {
  switch (protocol.toLowerCase()) {
    case 'http':
      return 'HTTP';
    case 'tcp':
      return 'TCP';
    default:
      return 'gRPC';
  }
};

export interface DestService {
  cluster: string;
  name: string;
  namespace: string;
}

export interface DestService {
  cluster: string;
  name: string;
  namespace: string;
}

export interface SEInfo {
  hosts: string[];
  location: string;
  namespace: string; // namespace represents where the ServiceEntry object is defined and not necessarily the namespace of the node.
}

export interface WEInfo {
  name: string;
}

export interface GraphRequestsHealth {
  healthAnnotations: { [idx: string]: string };
  inbound: { [idx: string]: { [idx: string]: number } };
  outbound: { [idx: string]: { [idx: string]: number } };
}

export interface GraphWorkloadStatus {
  availableReplicas: number;
  currentReplicas: number;
  desiredReplicas: number;
  name: string;
  syncedProxies: number;
}

export interface GraphNodeAppHealth {
  requests: GraphRequestsHealth;
  workloadStatuses: GraphWorkloadStatus[];
}

export interface GraphNodeWorkloadHealth {
  requests: GraphRequestsHealth;
  workloadStatus: GraphWorkloadStatus;
}

export interface GraphNodeServiceHealth {
  requests: GraphRequestsHealth;
}

export type GraphNodeHealthData = GraphNodeAppHealth | GraphNodeWorkloadHealth | GraphNodeServiceHealth | [] | null;

// Node data expected from server
export interface GraphNodeData {
  aggregate?: string;
  aggregateValue?: string;
  app?: string;
  cluster: string;
  destServices?: DestService[];
  hasCB?: boolean;
  hasFaultInjection?: boolean;
  hasHealthConfig?: HealthAnnotationType;
  hasMirroring?: boolean;
  hasRequestRouting?: boolean;
  hasRequestTimeout?: boolean;
  hasTCPTrafficShifting?: boolean;
  hasTrafficShifting?: boolean;
  hasVS?: {
    hostnames?: string[];
  };
  hasWorkloadEntry?: WEInfo[];
  healthData?: GraphNodeHealthData;
  id: string;
  isAmbient?: boolean;
  isBox?: string;
  isDead?: boolean;
  isGateway?: {
    egressInfo?: {
      hostnames?: string[];
    };
    gatewayAPIInfo?: {
      hostnames?: string[];
    };
    ingressInfo?: {
      hostnames?: string[];
    };
  };
  isIdle?: boolean;
  isInaccessible?: boolean;
  isK8sGatewayAPI?: boolean;
  isMisconfigured?: string;
  isOutOfMesh?: boolean;
  isOutside?: boolean;
  isRoot?: boolean;
  isServiceEntry?: SEInfo;
  isWaypoint?: boolean;
  labels?: { [key: string]: string };
  namespace: string;
  nodeType: NodeType;
  parent?: string;
  service?: string;
  traffic?: ProtocolTraffic[];
  version?: string;
  workload?: string;
}

// Edge data expected from server
export interface GraphEdgeData {
  destPrincipal?: string;
  id: string;
  isMTLS?: number;
  responseTime?: number;
  source: string;
  sourcePrincipal?: string;
  target: string;
  traffic?: ProtocolTraffic;
}

export interface GraphNodeWrapper {
  data: GraphNodeData;
}

export interface GraphEdgeWrapper {
  data: GraphEdgeData;
}

export interface GraphElements {
  edges?: GraphEdgeWrapper[];
  nodes?: GraphNodeWrapper[];
}

export interface GraphElementsQuery {
  appenders?: AppenderString;
  boxBy?: string;
  duration?: string;
  graphType?: GraphType;
  includeIdleEdges?: boolean;
  injectServiceNodes?: boolean;
  namespaces?: string;
  queryTime?: string;
  rateGrpc?: string;
  rateHttp?: string;
  rateTcp?: string;
  responseTime?: string;
  throughputType?: string;
  waypoints?: boolean;
}

export interface GraphDefinition {
  duration: number;
  elements: GraphElements;
  graphType: GraphType;
  timestamp: number;
}

// Node data after decorating at fetch-time (what is mainly used by ui code)
export interface DecoratedGraphNodeData extends GraphNodeData {
  grpcIn: number;
  grpcInErr: number;
  grpcInNoResponse: number;
  grpcOut: number;
  health: Health;
  healthStatus: string; // status name
  httpIn: number;
  httpIn3xx: number;
  httpIn4xx: number;
  httpIn5xx: number;
  httpInNoResponse: number;
  httpOut: number;
  // true if has istio namespace
  isIstio?: boolean;
  // assigned when node ranking is enabled. relative importance from most to least important [1..100]. Multiple nodes can have same rank.
  rank?: number;
  tcpIn: number;
  tcpOut: number;
  traffic: never;
}

// Edge data after decorating at fetch-time (what is mainly used by ui code)
export interface DecoratedGraphEdgeData extends GraphEdgeData {
  display: string;
  grpc: number;
  grpcErr: number;
  grpcNoResponse: number;
  grpcPercentErr: number;
  grpcPercentReq: number;
  // During the decoration process, we make non-optional some number attributes (giving them a default value)
  // computed, true if traffic rate > 0
  hasTraffic?: boolean;
  // assigned when graph is updated, the edge health depends on the node health, traffic, and config
  healthStatus?: string; // status name
  http: number;
  http3xx: number;
  http4xx: number;
  http5xx: number;
  httpNoResponse: number;
  httpPercentErr: number;
  httpPercentReq: number;
  // Default value -1
  isMTLS: number;
  protocol: ValidProtocols;
  // Default value NaN
  responseTime: number;
  responses: Responses;
  tcp: number;
  // Default value NaN
  throughput: number;
}

export interface DecoratedGraphNodeWrapper {
  data: DecoratedGraphNodeData;
}

export interface DecoratedGraphEdgeWrapper {
  data: DecoratedGraphEdgeData;
}

export interface DecoratedGraphElements {
  edges?: DecoratedGraphEdgeWrapper[];
  nodes?: DecoratedGraphNodeWrapper[];
}

export const EdgeAttr = {
  destPrincipal: 'destPrincipal',
  grpc: 'grpc',
  grpcErr: 'grpcErr',
  grpcNoResponse: 'grpcNoResponse',
  grpcPercentErr: 'grpcPercentErr',
  grpcPercentReq: 'grpcPercentReq',
  hasTraffic: 'hasTraffic',
  healthStatus: 'healthStatus',
  http: 'http',
  http3xx: 'http3xx',
  http4xx: 'http4xx',
  http5xx: 'http5xx',
  httpNoResponse: 'httpNoResponse',
  httpPercentErr: 'httpPercentErr',
  httpPercentReq: 'httpPercentReq',
  id: 'id',
  isMTLS: 'isMTLS',
  protocol: 'protocol',
  responses: 'responses',
  responseTime: 'responseTime',
  sourcePrincipal: 'sourcePrincipal',
  tcp: 'tcp',
  throughput: 'throughput'
};

export const NodeAttr = {
  aggregate: 'aggregate',
  aggregateValue: 'aggregateValue',
  app: 'app',
  cluster: 'cluster',
  destServices: 'destServices',
  grpcIn: 'grpcIn',
  grpcInErr: 'grpcInErr',
  grpcInNoResponse: 'grpcInNoResponse',
  grpcOut: 'grpcOut',
  hasCB: 'hasCB',
  hasFaultInjection: 'hasFaultInjection',
  hasMirroring: 'hasMirroring',
  hasRequestRouting: 'hasRequestRouting',
  hasRequestTimeout: 'hasRequestTimeout',
  hasTCPTrafficShifting: 'hasTCPTrafficShifting',
  hasTrafficShifting: 'hasTrafficShifting',
  hasVS: 'hasVS',
  hasWorkloadEntry: 'hasWorkloadEntry',
  health: 'health',
  healthStatus: 'healthStatus',
  httpIn: 'httpIn',
  httpIn3xx: 'httpIn3xx',
  httpIn4xx: 'httpIn4xx',
  httpIn5xx: 'httpIn5xx',
  httpInNoResponse: 'httpInNoResponse',
  httpOut: 'httpOut',
  id: 'id',
  isBox: 'isBox',
  isDead: 'isDead',
  isIdle: 'isIdle',
  isInaccessible: 'isInaccessible',
  isIstio: 'isIstio',
  isMisconfigured: 'isMisconfigured',
  isOutOfMesh: 'isOutOfMesh',
  isOutside: 'isOutside',
  isRoot: 'isRoot',
  isServiceEntry: 'isServiceEntry',
  isWaypoint: 'isWaypoint',
  namespace: 'namespace',
  nodeType: 'nodeType',
  rank: 'rank',
  service: 'service',
  tcpIn: 'tcpIn',
  tcpOut: 'tcpOut',
  version: 'version',
  workload: 'workload'
};
