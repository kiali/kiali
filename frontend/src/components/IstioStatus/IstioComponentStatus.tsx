import * as React from 'react';
import { ComponentStatus, Status, statusMsg } from '../../types/IstioStatus';
import { PFColors } from '../Pf/PfColors';
import {
  CheckCircleIcon,
  ExclamationCircleIcon,
  ExclamationTriangleIcon,
  MinusCircleIcon
} from '@patternfly/react-icons';
import { Split, SplitItem } from '@patternfly/react-core';
import { IconProps, createIcon } from 'config/KialiIcon';
import { kialiStyle } from 'styles/StyleUtils';
import { useKialiTranslation } from 'utils/I18nUtils';

type Props = {
  componentStatus: ComponentStatus;
};

const ErrorCoreComponent: IconProps = {
  color: PFColors.Danger,
  icon: ExclamationCircleIcon
};

const ErrorAddonComponent: IconProps = {
  color: PFColors.Warning,
  icon: ExclamationTriangleIcon
};

const NotReadyComponent: IconProps = {
  color: PFColors.Info,
  icon: MinusCircleIcon
};

const SuccessComponent: IconProps = {
  color: PFColors.Success,
  icon: CheckCircleIcon
};

// Mapping Valid-Core to Icon representation.
const validToIcon: { [valid: string]: IconProps } = {
  'false-false': ErrorAddonComponent,
  'false-true': ErrorCoreComponent,
  'true-false': SuccessComponent,
  'true-true': SuccessComponent
};

const splitItemStyle = kialiStyle({
  textAlign: 'left'
});

export const IstioComponentStatus: React.FC<Props> = (props: Props) => {
  const { t } = useKialiTranslation();

  const renderIcon = (status: Status, isCore: boolean): React.ReactNode => {
    let compIcon = validToIcon[`${status === Status.Healthy}-${isCore}`];

    if (status === Status.NotReady) {
      compIcon = NotReadyComponent;
    }

    compIcon.className = kialiStyle({
      marginTop: '0.25rem'
    });

    return createIcon(compIcon);
  };

  const renderCells = (): React.ReactNode => {
    const comp = props.componentStatus;

    return [
      <Split key={`cell-status-icon-${comp.name}`} hasGutter={true} className={splitItemStyle}>
        <SplitItem>{renderIcon(props.componentStatus.status, props.componentStatus.isCore)}</SplitItem>
        <SplitItem isFilled={true}>{comp.name}</SplitItem>
        <SplitItem>{t(statusMsg[comp.status])}</SplitItem>
      </Split>
    ];
  };

  return <>{renderCells()}</>;
};
