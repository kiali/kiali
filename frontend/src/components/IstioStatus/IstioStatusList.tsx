import * as React from 'react';
import { List } from '@patternfly/react-core';
import { ComponentStatus, Status } from '../../types/IstioStatus';
import { IstioComponentStatus } from './IstioComponentStatus';
import { kialiStyle } from 'styles/StyleUtils';
import { useKialiTranslation } from 'utils/I18nUtils';

type Props = {
  cluster: string;
  status: ComponentStatus[];
};

const listStyle = kialiStyle({
  paddingLeft: 0,
  marginTop: 0,
  marginLeft: 0
});

export const IstioStatusList: React.FC<Props> = (props: Props) => {
  const { t } = useKialiTranslation();

  const nonhealthyComponents = (): ComponentStatus[] => {
    return props.status.filter((c: ComponentStatus) => c.status !== Status.Healthy);
  };

  const coreComponentsStatus = (): ComponentStatus[] => {
    return nonhealthyComponents().filter((s: ComponentStatus) => s.is_core);
  };

  const addonComponentsStatus = (): ComponentStatus[] => {
    return nonhealthyComponents().filter((s: ComponentStatus) => !s.is_core);
  };

  const renderComponentList = (): React.ReactNode => {
    const groups = {
      core: coreComponentsStatus,
      addon: addonComponentsStatus
    };

    return ['core', 'addon'].map((group: string) => {
      return (
        <React.Fragment key={`status-${group}`}>
          {groups[group]().map((status: ComponentStatus) => {
            return <IstioComponentStatus key={`status-${group}-${status.name}`} componentStatus={status} />;
          })}
        </React.Fragment>
      );
    });
  };

  return (
    <List id="istio-status" aria-label={t('Istio Component List')} className={listStyle}>
      {renderComponentList()}
    </List>
  );
};
