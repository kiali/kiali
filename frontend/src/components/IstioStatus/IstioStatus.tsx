import * as React from 'react';
import { SVGIconProps } from '@patternfly/react-icons/dist/js/createIcon';
import * as API from '../../services/Api';
import * as AlertUtils from '../../utils/AlertUtils';
import { TimeInMilliseconds } from '../../types/Common';
import { ComponentStatus, Status, statusSeverity } from '../../types/IstioStatus';
import { MessageType } from '../../types/MessageCenter';
import { Namespace } from '../../types/Namespace';
import { KialiAppState } from '../../store/Store';
import { istioStatusSelector, namespaceItemsSelector } from '../../store/Selectors';
import { IstioStatusActions } from '../../actions/IstioStatusActions';
import { connect } from 'react-redux';
import { Text, TextVariants, TextContent, Tooltip, TooltipPosition, Label } from '@patternfly/react-core';
import { IstioStatusList } from './IstioStatusList';
import { PFColors } from '../Pf/PfColors';
import {
  CheckCircleIcon,
  ExclamationCircleIcon,
  ExclamationTriangleIcon,
  InfoCircleIcon,
  QuestionCircleIcon
} from '@patternfly/react-icons';
import { KialiDispatch } from 'types/Redux';
import { NamespaceThunkActions } from '../../actions/NamespaceThunkActions';
import { connectRefresh } from '../Refresh/connectRefresh';
import { kialiStyle } from 'styles/StyleUtils';
import { IconProps, createIcon, KialiIcon } from 'config/KialiIcon';
import { Link } from 'react-router-dom-v5-compat';
import { useKialiTranslation } from 'utils/I18nUtils';
import { MASTHEAD } from 'components/Nav/Masthead/Masthead';
import { isControlPlaneAccessible } from '../../utils/MeshUtils';
import { homeCluster } from '../../config';
import { PFBadge, PFBadges } from '../Pf/PfBadges';

export type ClusterStatusMap = { [cluster: string]: ComponentStatus[] };

type ReduxStateProps = {
  namespaces?: Namespace[];
  statusMap: ClusterStatusMap; // map of clusters to ComponentStatus[]
};

type ReduxDispatchProps = {
  refreshNamespaces: () => void;
  setIstioStatus: (statusMap: ClusterStatusMap) => void;
};

type StatusIcons = {
  ErrorIcon?: React.ComponentClass<SVGIconProps>;
  HealthyIcon?: React.ComponentClass<SVGIconProps>;
  InfoIcon?: React.ComponentClass<SVGIconProps>;
  WarningIcon?: React.ComponentClass<SVGIconProps>;
};

type Props = ReduxStateProps &
  ReduxDispatchProps & {
    icons?: StatusIcons;
    lastRefreshAt: TimeInMilliseconds;
    location?: string;
  };

const ValidToColor = {
  'true-true-true': PFColors.Danger,
  'true-true-false': PFColors.Danger,
  'true-false-true': PFColors.Danger,
  'true-false-false': PFColors.Danger,
  'false-true-true': PFColors.Warning,
  'false-true-false': PFColors.Warning,
  'false-false-true': PFColors.Info,
  'false-false-false': PFColors.Success
};

const defaultIcons = {
  ErrorIcon: ExclamationCircleIcon,
  HealthyIcon: CheckCircleIcon,
  InfoIcon: InfoCircleIcon,
  WarningIcon: ExclamationTriangleIcon
};

const iconStyle = kialiStyle({
  marginLeft: '0.5rem',
  fontSize: '1rem'
});

const clusterStyle = kialiStyle({
  display: 'flex',
  alignItems: 'center'
});

const labelStyle = kialiStyle({
  $nest: {
    '& .pf-v5-c-label__icon': {
      marginRight: '0.5rem'
    }
  }
});

export const meshLinkStyle = kialiStyle({
  display: 'flex',
  justifyContent: 'center',
  marginTop: '0.75rem',
  $nest: {
    '& > span': {
      marginRight: '0.5rem'
    }
  }
});

export const IstioStatusComponent: React.FC<Props> = (props: Props) => {
  const { t } = useKialiTranslation();

  const { namespaces, setIstioStatus, refreshNamespaces, lastRefreshAt } = props;

  React.useEffect(() => {
    refreshNamespaces();
  }, [refreshNamespaces]);

  const fetchStatus = React.useCallback((): void => {
    API.getIstioStatus()
      .then(response => {
        const statusMap: ClusterStatusMap = {};

        response.data.forEach(status => {
          if (!statusMap[status.cluster]) {
            statusMap[status.cluster] = [];
          }
          statusMap[status.cluster].push(status);
        });

        setIstioStatus(statusMap);
      })
      .catch(error => {
        // User without namespaces can't have access to mTLS information. Reduce severity to info.
        const informative = namespaces && namespaces.length < 1;

        if (informative) {
          AlertUtils.addError(t('Istio deployment status disabled.'), error, 'default', MessageType.INFO);
        } else {
          AlertUtils.addError(t('Error fetching Istio deployment status.'), error, 'default', MessageType.ERROR);
        }
      });
  }, [namespaces, setIstioStatus, t]);

  React.useEffect(() => {
    // retrieve status for all clusters
    fetchStatus();
  }, [lastRefreshAt, fetchStatus]);

  const getSeverity = (components: ComponentStatus[]): number =>
    Math.max(
      ...components.map(
        cs =>
          statusSeverity[cs.status] +
          (cs.is_core && statusSeverity[cs.status] !== statusSeverity[Status.Healthy] ? 10 : 0)
      )
    ); // non health core component has much higher severity

  const sortedClusters = Object.keys(props.statusMap).sort((a, b) => {
    const worstA = getSeverity(props.statusMap[a]);
    const worstB = getSeverity(props.statusMap[b]);
    return worstB - worstA;
  });

  const tooltipContent = (): React.ReactNode => {
    return (
      <>
        <TextContent style={{ color: PFColors.White }}>
          <Text component={TextVariants.h4}>{t('Cluster Status')}</Text>
          {sortedClusters.map(cl => (
            <>
              <div className={clusterStyle}>
                <PFBadge badge={PFBadges.Cluster} size="sm" />
                {cl}
                {cl === homeCluster?.name && (
                  <span style={{ marginLeft: '0.25rem' }}>
                    <KialiIcon.Star />
                  </span>
                )}
              </div>
              <IstioStatusList key={cl} status={props.statusMap[cl] || []} cluster={cl} />
            </>
          ))}
          {!props.location?.endsWith('/mesh') && isControlPlaneAccessible() && (
            <div className={meshLinkStyle}>
              <span>{t('More info at')}</span>
              <Link to="/mesh">{t('Mesh page')}</Link>
            </div>
          )}
        </TextContent>
      </>
    );
  };

  const tooltipColor = (): string => {
    let coreUnhealthy = false;
    let addonUnhealthy = false;
    let notReady = false;
    const values = Object.values(props.statusMap).flat();

    Object.keys(values ?? {}).forEach((compKey: string) => {
      const { status, is_core } = values[compKey];
      const isNotReady: boolean = status === Status.NotReady;
      const isUnhealthy: boolean = status !== Status.Healthy && !isNotReady;

      if (is_core) {
        coreUnhealthy = coreUnhealthy || isUnhealthy;
      } else {
        addonUnhealthy = addonUnhealthy || isUnhealthy;
      }

      notReady = notReady || isNotReady;
    });

    return ValidToColor[`${coreUnhealthy}-${addonUnhealthy}-${notReady}`];
  };

  const healthyComponents = (): boolean => {
    const values = Object.values(props.statusMap).flat();
    return values.reduce((healthy: boolean, compStatus: ComponentStatus) => {
      return healthy && compStatus.status === Status.Healthy;
    }, true);
  };

  const tooltipPosition = props.location === MASTHEAD ? TooltipPosition.bottom : TooltipPosition.top;

  let statusIcon: React.ReactElement;

  if (!healthyComponents()) {
    const icons = props.icons ? { ...defaultIcons, ...props.icons } : defaultIcons;
    const iconColor = tooltipColor();
    let icon = QuestionCircleIcon;
    let dataTest = 'istio-status';

    if (iconColor === PFColors.Danger) {
      icon = icons.ErrorIcon;
      dataTest = `${dataTest}-danger`;
    } else if (iconColor === PFColors.Warning) {
      icon = icons.WarningIcon;
      dataTest = `${dataTest}-warning`;
    } else if (iconColor === PFColors.Info) {
      icon = icons.InfoIcon;
      dataTest = `${dataTest}-info`;
    } else if (iconColor === PFColors.Success) {
      icon = icons.HealthyIcon;
      dataTest = `${dataTest}-success`;
    }

    const iconProps: IconProps = {
      className: iconStyle,
      dataTest: dataTest
    };

    statusIcon = createIcon(iconProps, icon, iconColor);
  } else {
    const iconProps: IconProps = {
      className: iconStyle,
      dataTest: 'istio-status-success'
    };

    statusIcon = createIcon(iconProps, defaultIcons.HealthyIcon, ValidToColor['false-false-false']);
  }

  return (
    <Tooltip
      data-test="component-status-tooltip"
      position={tooltipPosition}
      enableFlip={true}
      content={tooltipContent()}
      maxWidth="25rem"
    >
      <>
        {homeCluster?.name && (
          <Label className={labelStyle} data-test="cluster-icon" color="blue" icon={<KialiIcon.Cluster />}>
            {homeCluster?.name}
            {isControlPlaneAccessible() && statusIcon}
          </Label>
        )}
      </>
    </Tooltip>
  );
};

const mapStateToProps = (state: KialiAppState): ReduxStateProps => ({
  statusMap: istioStatusSelector(state),
  namespaces: namespaceItemsSelector(state)
});

const mapDispatchToProps = (dispatch: KialiDispatch): ReduxDispatchProps => ({
  setIstioStatus: (statusMap: ClusterStatusMap) => {
    dispatch(IstioStatusActions.setinfo(statusMap));
  },
  refreshNamespaces: () => {
    dispatch(NamespaceThunkActions.fetchNamespacesIfNeeded());
  }
});

export const IstioStatus = connectRefresh(connect(mapStateToProps, mapDispatchToProps)(IstioStatusComponent));
