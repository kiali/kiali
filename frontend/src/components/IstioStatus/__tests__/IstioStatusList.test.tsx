import * as React from 'react';
import { shallow } from 'enzyme';
import { ComponentStatus, Status } from '../../../types/IstioStatus';
import { IstioStatusList } from '../IstioStatusList';
import { shallowToJson } from 'enzyme-to-json';
import { CLUSTER_DEFAULT } from '../../../types/Graph';

it('lists all the components grouped', () => {
  const components: ComponentStatus[] = [
    {
      cluster: CLUSTER_DEFAULT,
      name: 'grafana',
      status: Status.NotFound,
      is_core: false
    },
    {
      cluster: CLUSTER_DEFAULT,
      name: 'prometheus',
      status: Status.Unhealthy,
      is_core: false
    },
    {
      cluster: CLUSTER_DEFAULT,
      name: 'istiod',
      status: Status.NotFound,
      is_core: true
    },
    {
      cluster: CLUSTER_DEFAULT,
      name: 'istio-egressgateway',
      status: Status.Unhealthy,
      is_core: true
    }
  ];

  const wrapper = shallow(<IstioStatusList status={components} cluster={'Kubernetes'} />);

  expect(shallowToJson(wrapper)).toBeDefined();
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});
