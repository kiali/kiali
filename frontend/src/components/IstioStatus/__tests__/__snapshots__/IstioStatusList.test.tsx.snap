// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lists all the components grouped 1`] = `
<List
  aria-label="Istio Component List"
  className="kiali_f6tdqla"
  id="istio-status"
>
  <IstioComponentStatus
    componentStatus={
      Object {
        "cluster": "Kubernetes",
        "is_core": true,
        "name": "istiod",
        "status": "NotFound",
      }
    }
    key="status-core-istiod"
  />
  <IstioComponentStatus
    componentStatus={
      Object {
        "cluster": "Kubernetes",
        "is_core": true,
        "name": "istio-egressgateway",
        "status": "Unhealthy",
      }
    }
    key="status-core-istio-egressgateway"
  />
  <IstioComponentStatus
    componentStatus={
      Object {
        "cluster": "Kubernetes",
        "is_core": false,
        "name": "grafana",
        "status": "NotFound",
      }
    }
    key="status-addon-grafana"
  />
  <IstioComponentStatus
    componentStatus={
      Object {
        "cluster": "Kubernetes",
        "is_core": false,
        "name": "prometheus",
        "status": "Unhealthy",
      }
    }
    key="status-addon-prometheus"
  />
</List>
`;
